File: emulinker/src/main/java/org/emulinker/kaillera/controller/messaging/ByteBufferMessage.java
package org.emulinker.kaillera.controller.messaging;

import java.nio.Buffer;
import java.nio.ByteBuffer;

public abstract class ByteBufferMessage {
  private ByteBuffer buffer;

  public abstract int getLength();

  public abstract String description();

  @Override
  public abstract String toString();

  protected void initBuffer() {
    initBuffer(getLength());
  }

  private void initBuffer(int size) {
    buffer = getBuffer(size);
  }

  public void releaseBuffer() {}

  public ByteBuffer toBuffer() {
    initBuffer();
    writeTo(buffer);
    // Cast to avoid issue with java version mismatch: https://stackoverflow.com/a/61267496/2875073
    ((Buffer) buffer).flip();
    return buffer;
  }

  public abstract void writeTo(ByteBuffer buffer);

  public static ByteBuffer getBuffer(int size) {
    return ByteBuffer.allocateDirect(size);
  }

  public static void releaseBuffer(ByteBuffer buffer) {
    // nothing to do since we aren't caching buffers anymore
  }
}

File: emulinker/src/main/java/org/emulinker/kaillera/controller/messaging/ByteBufferMessage.kt
package org.emulinker.kaillera.controller.messaging

import java.nio.Buffer
import java.nio.ByteBuffer

abstract class ByteBufferMessage {
  private lateinit var buffer: ByteBuffer

  abstract val length: Int

  private fun initBuffer() {
    initBuffer(length)
  }

  private fun initBuffer(size: Int) {
    buffer = getBuffer(size)
  }

  fun releaseBuffer() {}
  fun toBuffer(): ByteBuffer {
    initBuffer()
    writeTo(buffer)
    // Cast to avoid issue with java version mismatch: https://stackoverflow.com/a/61267496/2875073
    (buffer as Buffer).flip()
    return buffer
  }

  abstract fun writeTo(buffer: ByteBuffer)

  companion object {
    fun getBuffer(size: Int): ByteBuffer {
      return ByteBuffer.allocateDirect(size)
    }
  }
}
